### PR TITLE
WIP: Rounding errors at build overviews module

### PR DESCRIPTION
**Pull request by drons**
In the build overviews module, there are problems associated with rounding.

In the test, an image of the chessboard is generated, and then an overview image is created for it. 
Then I take an overview image and upsample it to the original size. 
For the chessboard, the images must fully match.

If the dimensions of the original image are NOT a multiple of overview decimation factors then generated overview contains unexpected pixels.

The rounding error arises because we re-calculate the decimation factor form integer image sizes in the function GDALRegenerateOverviews (dfXRatioDstToSrc and dfYRatioDstToSrc values).
These rounding errors lead to problems similar to this one (https://trac.osgeo.org/gdal/ticket/7020), spotted out by @mojodna.

The solution can be rounding the the decimation factors values at GDALRegenerateOverviews or explicitly passing them to a function.

@rouault Which solution is better to choose?

I think the first is easier and faster. In functions of the upper level, decimation factors are represented as integers and we do not violate anything.

### CODE REVIEW

**Comment by drons** [comment]
I'll try to explain more.

This is original image of size 259x259:
original

I am create overviews
`ds.BuildOverviews('AVERAGE', overviewlist = [2, 4, 8, 16])`
Read and upsample top overview image. Result is incredible:
upsampled_overview
But expected result must match original image.

What is happening at this point in the code?
* GDALDataset::BuildOverviews takes integer list of overview decimation factors and pass it to GDALDefaultOverviews::BuildOverviews and then to GTIFFBuildOverviews
* At [GTIFFBuildOverviews](https://github.com/OSGeo/gdal/blob/master/gdal/frmts/gtiff/gt_overview.cpp#L607) we compute overview images size. Size is an integer value! For overview with factor 2 we have:  `nOXSize = (259 + 2 - 1) / 2 = 130`
* Go deeper. [Here](https://github.com/OSGeo/gdal/blob/master/gdal/gcore/overview.cpp#L3096) GDALRegenerateOverviews compute decimation factors again from image sizes. Decimation factor is float here! `dfXRatioDstToSrc = static_cast<double>(259) / 130 = 1.992307692`
* Incorrect decimation factors lead to incorrectly computed weights of the image convolution at GDALResampleChunk*.

**Comment by drons** [comment]
This problem can be reproduced using third-party programs. For example Erdas Imagine.
You need to select an image, copy it to two different folders and build an overviews using gdaladdo and using another program.

Overlay both images in one window and run flicker tool.
At overview scales you can see picture like this:
flicker

**Comment by rouault** [comment]
dfXRatioDstToSrc = static_cast<double>(259) / 130 = 1.992307692 is technically correct. The artifacts you get then are expected. I'd think that if you tweak the ratio to be 2 you could get a slightly displacement of one pixel at the right or bottom of the image. Or we have perhaps a sampling issue in the resampling kernel itself.
This is an area of the code super tricky and sensitive, so I'm super cautious with changes there. Some artifacts shown in screenshots are sometimes due to the display tool doing itself resampling (in a ticket I found the issue was due to how QGIS interacted with GDAL for example), or not properly taking into account the georeferencing of the overview.
In any case, the unconditional rounding you added in 3b6d2532827ac13f1064e54925a165c9e74922d2 is certainly incorrect in case someone would have manually constructed an overview level with a factor of 2.5 for example.

**Comment by drons** [comment]
> dfXRatioDstToSrc = static_cast(259) / 130 = 1.992307692 is technically correct.
Certainly. But the expected value is 2. 2 was passed to GDALDataset::BuildOverviews.
If I reduce the image by 2 times by the average algorithm, then I mean that the pixel of the overview image will be strictly equal to the average value of the four underlying pixels.
With a dfXRatioDstToSrc = 1.992307692, it is not true.

>  I'd think that if you tweak the ratio to be 2 you could get a slightly displacement of one pixel at the right or bottom of the image.
On the contrary, this removes the 'visual' displacements.

>This is an area of the code super tricky and sensitive, so I'm super cautious with changes there.
Yes. I see. Some tests that use overview images faiied. I think that we need manually review its results.

>Some artifacts shown in screenshots are sometimes due to the display tool doing itself resampling (in a >ticket I found the issue was due to how QGIS interacted with GDAL for example), or not properly taking >into account the georeferencing of the overview.
I'm test it in Erdas. We will see the artifacts if overlay the image downsampled with the Erdas reproject tool onto the image with the overviews built with gdaladdo.
May be it can be reproduced with gdaladdo and gdalwarp too. It have a different sampling pipeline. I will test later.

>In any case, the unconditional rounding you added in 3b6d253 is certainly incorrect in case someone >would have manually constructed an overview level with a factor of 2.5 for example.
It's unusual scale factor. But a real case.
Maybe I'll try the second solution?
Explicitly pass values at GDALRegenerateOverviews.
But it break ABI.

**Comment by drons** [comment]
I think we can not neglect the most frequent usage for the sake of rare and extreme cases.
> That would be annoying.

May be add GDALRegenerateOverviewsInternal with detimation factor parameters? Use it only in our library, and update GDALRegenerateOverviews on next major release.

**Comment by drons** [comment]
@rouault  I'm compare 'warp' and 'overview' pipeline

Function `downsample_warp_vs_translate` generate several images:
* Test dataset - 10117x5578_chessboard_3.tif
* Downsample via reproject - 10117x5578_downsample_via_reproject.tif
* Downsample via translate without overviews - 10117x5578_downsample_via_translate_no_ov.tif
* Downsample via translate with overviews - 10117x5578_downsample_via_translate_with_ov.tif
* Dumped toplevel overview - 10117x5578_overview_dump.tif

After that, using the ImageMagic tool, gif-animation with image pairs were сreated:
```
convert -delay 100 -page +0+0 10117x5578_downsample_via_reproject.tif -page +0+0 10117x5578_downsample_via_translate_no_ov.tif -loop 0 ./no_ov.gif
```
no_ov
There is no any flicker between reprojected image vs downsampled without overviews 

```
convert -delay 100 -page +0+0 10117x5578_downsample_via_reproject.tif -page +0+0 10117x5578_downsample_via_translate_with_ov.tif -loop 0 ./with_ov.gif
```
with_ov
It is clear that some pixels are not only shifted, but also lost if we use overviews while downsampling.

```
convert -delay 100 -page +0+0 10117x5578_downsample_via_reproject.tif -page +0+0 10117x5578_overview_dump.tif -loop 0 ./overview.gif
```

The result of the last comparison is attached to the message due to GH problems. [results.zip](https://github.com/OSGeo/gdal/files/1958101/results.zip)

The saddest thing is that the results of downsampling via translate with overviews and overview dump is differ. This suggests that the rounding problems in the RasterIO too.

**Comment by drons** [comment]
If I add rounding to the GDALRegenerateOverviews function (as at 3b6d2532827ac13f1064e54925a165c9e74922d2), then the edges of the cells become without artifacts, but are still shifted:

with_ov

**Comment by drons** [comment]
I'm add "OVERVIEW_FACTOR" metadata value to "GDAL_INTERNAL" domain, and use it inside GDALRegenerateOverviews to compute actual "dst to src" ratio.

Moreover, these metadata can be used to calculate actual overview level inside GDALBandGetBestOverviewLevel2 function in RasterIO code.

@rouault If this is the right direction, then I will add this change to the rest of the code.

**Comment by rouault** [comment]
> @rouault If this is the right direction, then I will add this change to the rest of the code.

I'm not convinced there's an issue at all... Those artifacts you see are perhaps/probably just a mathematical consequence of the overview_pixel_width not being an exact value of baseraster_pixel_width / overview_factor in the case where baseraster_pixel_width is a odd number and overview_factor = 2. Or there might be an issue in the resampling kernel. But all that hacking around modifying dfXRatioDstToSrc doesn't seem to solve anything.

What I believe is that you can't get both 'exact' spatial registration of the overview and artifact-free resampling on chessboard image. I'd hope GDAL to implement the 'exact' spatial registration case, and if that's not the case, that should be fixed, (with the consequence of artifacts on the chessboard images). Actually, I believe that with AVERAGE you can't get  'exact' spatial registration because, as soon as a source pixel is selected, its value is not weighted by some distance (similarly to NEAREST). You'd probably get better results with BILINEAR (or CUBIC, CUBICSPLINE, LANCZOS), which all use convolution filters.

**Comment by drons** [comment]
> Those artifacts you see are perhaps/probably just a mathematical consequence of the overview_pixel_width not being an exact value of baseraster_pixel_width / overview_factor in the case where baseraster_pixel_width is a odd number and overview_factor = 2.

The problem is still there.

> But all that hacking around modifying dfXRatioDstToSrc doesn't seem to solve anything.

On the contrary, the `upsample_overview_2` test shows that the library's behavior has become expected.
Or maybe there is an error in the test?

> What I believe is that you can't get both 'exact' spatial registration of the overview and artifact-free resampling on chessboard image.

It is possible with GDAL warp engine with AVERAGE sampling function, see `downsample_warp_vs_translate` `# Run reproject tool` section. 
I checked reprojected image in third-party software. Moreover, it can be seen that all cells are the same size and there are no artifacts on the edges. Also, I see that the result of reprojection has no problems with spatial registration. And it can be taken as a 'expected behavior'. 

But result of gdal_translate is differ from result of reproject if we run translate from image with overviews:
with_ov
According to my expectations, the results should be the same.

But an explicit overview factor pass to GDALRegenerateOverviews function of  solves only half of the problem. Only clear edge artifacts. The second half can be solved by passing the overview factor to RasterIO functions

I will try the remaining overview algorithms (BILINEAR, CUBIC, CUBICSPLINE, LANCZOS) some later.

I understand that the problem is too complicated and its solution can affect many tests. Maybe we should involve the community for discussion?

File: autotest/gcore/upsample_overview.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                   format=drv_name, xRes=scale, yRes=scale)
+
+    # Build overviews
+    src_ds = gdal.Open(src_name)
+    src_ds.BuildOverviews('AVERAGE', overviewlist = [2, 4, 8, 16, 32, 64])
+    src_ds = None
+
+    # Run Translate with overviews
+    translate_with_ov_ds_name = prefix + 'downsample_via_translate_with_ov.tif'
+    gdal.Translate(translate_with_ov_ds_name, src_name,
```
[the remark this answers is not in the document]
Comment (drons) [reply autotest/gcore/upsample_overview.py]: I expect that we must read exactly from the overview image. We must have overviews `[2, 4, 8, 16, 32, 64]` and read at scale 1/64

[the remark this answers is not in the document]
Comment (drons) [reply autotest/gcore/upsample_overview.py]: Moreover, In the absence of overview images, we get the expected result.

[the remark this answers is not in the document]
Comment (mojodna) [reply autotest/gcore/upsample_overview.py]: @drons are you using tiled overviews? if so, do you get the expected result when using regular overviews?

**Comment by drons** [comment]
Comparision result of warped image with translated (overview_alg='AVERAGE', translate_alg='BILINEAR')
with_ov_bilinear

**Comment by drons** [comment]
Comparision result of warped image with translated (overview_alg='BILINEAR', translate_alg='BILINEAR')
with_ov_bilinear_bilinear

Both combinations of algorithms produce an unexpected result.

File: autotest/gcore/upsample_overview.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (drons) [reply autotest/gcore/upsample_overview.py]: I'm use default overviews case. It is always tiled. At GTIFFBuildOverviews always called GTIFFWriteDirectory with arg bTiled=TRUE. There are no other cases. 
Only if hardcode GTIFFWriteDirectory call with bTiled=TRUE.  
In GDALRegenerateOverviews function, the processing is done by horizontal strips, not by tiles.

[the remark this answers is not in the document]
Comment (drons) [reply autotest/gcore/upsample_overview.py]: bTiled=FALSE did not affect anything. Overwiew written by strips, but result is the same.
```
#tiffinfo ./10117x5578_BILINEAR_BILINEAR_chessboard_3.tif.ovr 
TIFFReadDirectory: Warning, Unknown field with tag 42112 (0xa480) encountered.
TIFF Directory at offset 0x8 (8)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 5059 Image Length: 2789
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
  Tag 42112: <GDALMetadata><Item name="OVERVIEW_FACTOR" domain="GDAL_INTERNAL">2</Item></GDALMetadata>
TIFFReadDirectory: Warning, Unknown field with tag 42112 (0xa480) encountered.
TIFF Directory at offset 0x1b4 (436)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 2530 Image Length: 1395
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
  Tag 42112: <GDALMetadata><Item name="OVERVIEW_FACTOR" domain="GDAL_INTERNAL">4</Item></GDALMetadata>
TIFFReadDirectory: Warning, Unknown field with tag 42112 (0xa480) encountered.
TIFF Directory at offset 0x308 (776)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 1265 Image Length: 698
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
  Tag 42112: <GDALMetadata><Item name="OVERVIEW_FACTOR" domain="GDAL_INTERNAL">8</Item></GDALMetadata>
TIFFReadDirectory: Warning, Unknown field with tag 42112 (0xa480) encountered.
TIFF Directory at offset 0x434 (1076)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 633 Image Length: 349
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
  Tag 42112: <GDALMetadata><Item name="OVERVIEW_FACTOR" domain="GDAL_INTERNAL">16</Item></GDALMetadata>
TIFFReadDirectory: Warning, Unknown field with tag 42112 (0xa480) encountered.
TIFF Directory at offset 0x54a (1354)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 317 Image Length: 175
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
  Tag 42112: <GDALMetadata><Item name="OVERVIEW_FACTOR" domain="GDAL_INTERNAL">32</Item></GDALMetadata>
TIFFReadDirectory: Warning, Unknown field with tag 42112 (0xa480) encountered.
TIFFReadDirectory: Warning, Bogus "StripByteCounts" field, ignoring and calculating from imagelength.
TIFF Directory at offset 0x658 (1624)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 159 Image Length: 88
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Rows/Strip: 128
  Planar Configuration: single image plane
  Tag 42112: <GDALMetadata><Item name="OVERVIEW_FACTOR" domain="GDAL_INTERNAL">64</Item></GDALMetadata>
```

**Comment by drons** [comment]
@rouault 
>You'd probably get better results with BILINEAR (or CUBIC, CUBICSPLINE, LANCZOS), which all use convolution filters.

Using a BILINEAR, CUBIC, CUBICSPLINE, LANCZOS filters to build an overview image is not a good idea. Since these filters are not downsampling filters. Their use leads to loss of detail in the image. Good downsampling filters are AVERAGE and GAUSS.

From: https://en.wikipedia.org/wiki/Image_scaling
**Box sampling**
One weakness of bilinear, bicubic and related algorithms is that they sample a specific number of pixels. When down scaling below a certain threshold, such as more than twice for all bi-sampling algorithms, the algorithms will sample non-adjacent pixels, which results in both losing data, and causes rough results.

The trivial solution to this issue is box sampling, which is to consider the target pixel a box on the original image, and sample all pixels inside the box. This ensures that all input pixels contribute to the output. The major weakness of this algorithm is that it is hard to optimize.

But despite all this.
My task is simple. I need the overview images to be built with the 'overview factors' from user input, not the newly calculated ones based on the image and overview sizes.

Moreother, some raster drivers expect that, overview factors are only powers of two (see [GeoRasterDataset::IBuildOverviews](https://github.com/OSGeo/gdal/blob/master/gdal/frmts/georaster/georaster_dataset.cpp#L2765)), overview level is explicitly passed to [HFABand::CreateOverview](https://github.com/OSGeo/gdal/blob/master/gdal/frmts/hfa/hfaband.cpp#L2107). I'm think, that we must pass overview level to GDALRegenerateOverviews to use it directly.

**Comment by rouault** [comment]
> Using a BILINEAR, CUBIC, CUBICSPLINE, LANCZOS filters to build an overview image is not a good idea.

I believe the implementation in GDAL uses what they call 'box sampling'. All the input pixels will be taken into account.

> My task is simple. I need the overview images to be built with the 'overview factors' from user input, not the newly calculated ones based on the image and overview sizes.

I'm not sure that is what you want/need. Let consider, for the sake of changing of point of view, a that  IBuildOverviews() would take the width x hight values of each overview level instead of the overview factor. What would you do ?
And take into account my example of an input image of just 5x5 pixels (thus with an overview of 3x3): you'll see that using the overview factor of 2 instead of 5/3 will lead to spatial shifts in the overview.
As I said above I think AVERAGE (or its current implementation) will necessarily lead to those not-so-good result you observe as soon as you don't have dimensions whose ratio is integer overview factors. Probably that a WEIGHTED_AVERAGE variant that would take into account the proportion of the source pixel that hit the area of the target pixel would improve that.

**Comment by drons** [comment]
>I believe the implementation in GDAL uses what they call 'box sampling'. All the input pixels will be taken into account.

Implementation in GDAL is good, we just need to call with a few other parameters of RatioDstToSrc.

Maybe I should add a switch via CPLSetConfigOption between the old(default) (compute RatioDstToSrc from image dimensions) and the new approach (pass RatioDstToSrc from user input)?

**Comment by drons** [comment]
@rouault 
> Let consider, for the sake of changing of point of view, that IBuildOverviews() would take the width x hight values of each overview level instead of the overview factor. What would you do ?

I would continue to seek a solution to the problem.

> And take into account my example of an input image of just 5x5 pixels (thus with an overview of 3x3): you'll see that using the overview factor of 2 instead of 5/3 will lead to spatial shifts in the overview.

I think that it is not possible to consider the spatial registration of an overview image in isolation from the main image. If we know exactly the overview factor, then we can accurately restore the spatial registration without shifts.

> But all that hacking around modifying dfXRatioDstToSrc doesn't seem to solve anything.

I commented out warp-translate comparision in the test and added tests with GDAL_STRICT_OVERVIEW_FACTOR config option. For now if GDAL_STRICT_OVERVIEW_FACTOR=NO we have default GDAL overview behaviour (compute RatioDstToSrc from image dimensions, see test `upsample_overview_2a`), and if  GDAL_STRICT_OVERVIEW_FACTOR=YES we turn on new behaviour (compute RatioDstToSrc from image overview factors, see test `upsample_overview_2b`).
If we activate GDAL_STRICT_OVERVIEW_ACTOR then the test passes.

**Comment by rouault** [comment]
I'm not sure there is a value of letting this PR open given there is disagreement there is actually an issue

**Comment by drons** [comment]
I'm think PR can be closed.
I'm fix this problem at my side.

**Pull request closed without merging**